action/protocol/staking/candidate_center_test.go

```diff
@@ -461,59 +462,106 @@ func TestMultipleNonStakingCandidate(t *testing.T) {
+		Name:               "non-self-staked1-col-opt",
+		Votes:              big.NewInt(0),
+		SelfStakeBucketIdx: candidateNoSelfStakeBucketIndex,
+		SelfStake:          big.NewInt(0),
+	}
+	candNonStaked1ColName := &Candidate{
+		Owner:              identityset.Address(21),
+		Operator:           identityset.Address(23),
+		Reward:             identityset.Address(3),
 		Name:               "non-self-staked1",
 		Votes:              big.NewInt(0),
 		SelfStakeBucketIdx: candidateNoSelfStakeBucketIndex,
 		SelfStake:          big.NewInt(0),
 	}
+	candStakedColBucket := &Candidate{
+		Owner:              identityset.Address(4),
+		Operator:           identityset.Address(14),
+		Reward:             identityset.Address(3),
+		Name:               "self-staked-col-bucket",
+		Votes:              unit.ConvertIotxToRau(1200000),
+		SelfStakeBucketIdx: 1,
+		SelfStake:          unit.ConvertIotxToRau(1200000),
+	}
+
 	checkCandidates := func(candcenter *CandidateCenter, cands []*Candidate) {
 		r.True(testEqual(candcenter, CandidateList(cands)))
+		// commit
+		r.NoError(candcenter.Commit())
+		r.True(testEqual(candcenter, CandidateList(cands)))
+		// from state manager
+		dk := protocol.NewDock()
+		view := protocol.View{}
+		r.NoError(view.Write(_protocolID, candcenter))
+		dk.Reset()
+		candcenter = candCenterFromNewCandidateStateManager(r, view, dk)
+		r.True(testEqual(candcenter, CandidateList(cands)))
 	}
-	// add candidates
-	r.NoError(candcenter.Upsert(cand))
-	checkCandidates(candcenter, []*Candidate{cand})
-	// self stake candidate
-	candStaked := cand.Clone()
-	candStaked.SelfStakeBucketIdx = 1
-	candStaked.SelfStake = unit.ConvertIotxToRau(1200000)
-	r.NoError(candcenter.Upsert(candStaked))
-	checkCandidates(candcenter, []*Candidate{candStaked})
-	// change self stake bucket index
-	candUpdated := candStaked.Clone()
-	candUpdated.SelfStakeBucketIdx = 2
-	r.NoError(candcenter.Upsert(candUpdated))
-	checkCandidates(candcenter, []*Candidate{candUpdated})
-	r.False(candcenter.ContainsSelfStakingBucket(candStaked.SelfStakeBucketIdx))
+	t.Run("nonstaked candidate not collision on bucket", func(t *testing.T) {
+		candcenter, err := NewCandidateCenter(nil)
+		r.NoError(err)
+		r.NoError(candcenter.Upsert(candNonStaked1))
+		r.NoError(candcenter.Upsert(candNonStaked2))
+		checkCandidates(candcenter, []*Candidate{candNonStaked1, candNonStaked2})
+	})
+	t.Run("staked candidate collision on bucket", func(t *testing.T) {
+		candcenter, err := NewCandidateCenter(nil)
+		r.NoError(err)
+		r.NoError(candcenter.Upsert(candStaked))
+		r.ErrorIs(candcenter.Upsert(candStakedColBucket), ErrInvalidSelfStkIndex)
+		checkCandidates(candcenter, []*Candidate{candStaked})
+	})
+	t.Run("nonstaked candidate collision on operator", func(t *testing.T) {
+		candcenter, err := NewCandidateCenter(nil)
+		r.NoError(err)
+		r.NoError(candcenter.Upsert(candNonStaked1))
+		r.ErrorIs(candcenter.Upsert(candNonStaked1ColOpt), ErrInvalidOperator)
```

Comment:
can combine the next test into here by just adding 1 line:
`r.ErrorIs(candcenter.Upsert(candNonStaked1ColName), action.ErrInvalidCanName)`